### DISCUSSION TITLE
Code contribution

### DISCUSSION THREAD

**Discussion by jpelletier** [wwarthen/RomWBW#299] [Show and tell]
I ported a RTC driver to the Maxim DS1501/DS1511.

It currently misses the functions for reading/writing the NVRAM of the chip. I don't know what are the parameters for these functions.
Otherwise, it would be pretty much complete.
My code is attached below.
[ds1501rtc-asm.txt](https://github.com/wwarthen/RomWBW/files/9175047/ds1501rtc-asm.txt)
.

**Comment by wwarthen** [comment]
Very nice.  Thank you so much for the contribution.  I will integrate your driver in the next day or two.

Thank you!

Wayne

**Comment by wwarthen** [comment]
OK, I have attempted to integrate your driver into RomWBW.  It assembles
OK, but I have no way to test it.  If possible, perhaps you could try to
test it.  I have committed the driver integration to the GitHub repository.

By default the driver is not enabled because RomWBW HBIOS only supports a
single RTC enabled at one time.  The common DSRTC driver is the one that is
enabled by default.  To enable DS1501RTC, you will need to set DSRTCENABLE
to FALSE and set DS1501RTCENABLE to TRUE.

The driver needs a base address (RTC1501RTC_BASE).  I have added a config
setting for that and defaulted it to $50.  I had no reason for choosing
$50, I just needed to assign some value.  Let me know what is the best
value to use as the base address.

I added a call to RTC_SETDISP which is used to make it dispatchable by
RomWBW HBIOS.

I have also converted the IN0 and OUT0 instructions to IN and OUT
instructions.  This will allow it to run on Z80 systems.  This should still
be fine on a Z180 system as long as your Z180 system is not specifically
decoding the high 8 bits of the I/O port addresses (which is not normally
done).

Let me know if you see any issues with what I have done.

Thanks!

Wayne

**Comment by jpelletier** [comment]
I just added some code to access the NVM:

;
; NVRAM FUNCTIONS
;
DS1501RTC_GETBYT:
; C		Index
; E		Value
	; Set address
	LD	B, C
	LD	C, DS1501RTC_RAMADDR
	OUT (C), B

	; Get data
	IN	A, (DS1501RTC_RAMDATA)
	LD	E,A

	;Return success
	XOR	A
	RET

DS1501RTC_SETBYT:
; C		Index
; E		Value
	; Set address
	LD	B, C
	LD	C, DS1501RTC_RAMADDR
	OUT (C), B

	; Set data
	LD	A,E
	OUT	(DS1501RTC_RAMDATA), A

	;Return success
	XOR	A
	RET

DS1501RTC_GETBLK:
; HL	Buffer Address
	LD	B, 0	;256 Bytes

	;Set BME
	IN	A, (DS1501RTC_CONTROLB)
	OR	A, DS1501RTC_BME
	OUT	(DS1501RTC_CONTROLB), A

	XOR	A
	OUT	(DS1501RTC_RAMADDR), A
	LD	C, DS1501RTC_RAMDATA
	INIR

    ;Clear BME
	IN	A, (DS1501RTC_CONTROLB)
	AND	A, ~DS1501RTC_BME
	OUT	(DS1501RTC_CONTROLB), A

	;Return success
	XOR	A
	RET

DS1501RTC_SETBLK:
; HL	Buffer Address
	LD	B, 0	;256 Bytes

	;Set BME
	IN	A, (DS1501RTC_CONTROLB)
	OR	A, DS1501RTC_BME
	OUT	(DS1501RTC_CONTROLB), A

	XOR	A
	OUT	(DS1501RTC_RAMADDR), A
	LD	C, DS1501RTC_RAMDATA
	OTIR

    ;Clear BME
	IN	A, (DS1501RTC_CONTROLB)
	AND	A, ~DS1501RTC_BME
	OUT	(DS1501RTC_CONTROLB), A

	;Return success
	XOR	A
	RET

**Comment by wwarthen** [comment]
Great, thank you.  I will integrate this code in the next couple days.

Thanks,

Wayne

**Comment by wwarthen** [comment]
I have added this code in commit e286a42.

Thanks!

Wayne
